libs/langchain/langchain/chat_models/fake.py

```diff
@@ -16,28 +16,77 @@
 class FakeMessagesListChatModel(BaseChatModel):
     """Fake ChatModel for testing purposes."""
 
-    responses: List[BaseMessage]
+    responses: Union[List[BaseMessage], List[List[BaseMessage]]]
     sleep: Optional[float] = None
     i: int = 0
 
+    @property
+    def _llm_type(self) -> str:
+        return "fake-messages-list-chat-model"
+
+    @property
+    def _identifying_params(self) -> Dict[str, Any]:
+        return {"responses": self.responses}
+
     def _generate(
         self,
         messages: List[BaseMessage],
         stop: Optional[List[str]] = None,
         run_manager: Optional[CallbackManagerForLLMRun] = None,
         **kwargs: Any,
     ) -> ChatResult:
+        response = self._call(messages, stop=stop, run_manager=run_manager, **kwargs)
+        generation = ChatGeneration(message=response)
+        return ChatResult(generations=[generation])
+
+    def _call(
+        self,
+        messages: List[BaseMessage],
+        stop: Optional[List[str]] = None,
+        run_manager: Optional[CallbackManagerForLLMRun] = None,
+        **kwargs: Any,
+    ) -> Union[BaseMessage, List[BaseMessage]]:
+        """First try to lookup in queries, else return 'foo' or 'bar'."""
         response = self.responses[self.i]
         if self.i < len(self.responses) - 1:
             self.i += 1
         else:
             self.i = 0
-        generation = ChatGeneration(message=response)
-        return ChatResult(generations=[generation])
+        return response
 
-    @property
-    def _llm_type(self) -> str:
-        return "fake-messages-list-chat-model"
+    def _stream(
+        self,
+        messages: List[BaseMessage],
+        stop: Union[List[str], None] = None,
+        run_manager: Union[CallbackManagerForLLMRun, None] = None,
+        **kwargs: Any,
+    ) -> Iterator[ChatGenerationChunk]:
+        response = self.responses[self.i]
+        if self.i < len(self.responses) - 1:
+            self.i += 1
+        else:
+            self.i = 0
+        for c in response:
```

Comment:
again, not sure this gracefully handles both types for elements of self.responses.

Reply:
Unless I'm misunderstanding, I believe it does, this also works in my tests:

```python
    function_responses = [
        AIMessage(
            content="",
            additional_kwargs={
                "function_call": {
                    "name": "test_function",
                    "arguments": '{\n  "word": "foo",\n  "repeats": 2\n}',
                }
            },
        ),
        "Foo repeated twice is: foo foo",
    ]
```